cocos/3d/assets/animation-clip.js

```diff
@@ -33,387 +33,366 @@ import { clamp } from "../../core/vmath/utils";
-          }
-
-          if (jointFrames.scales) {
-            let a = jointFrames.scales[loIdx];
-            let b = jointFrames.scales[hiIdx];
-
-            vec3.lerp(tmpvec3, a, b, ratio);
-            jointState.blendScale(tmpvec3, weight);
-          }
-        }
-      }
+    get times() {
+        return this._times;
     }
-  }
 }
+cc.AnimationFrame = AnimationFrame;
 
-/**
- * The SamplingState class represents the progress of blended sampling.
- */
-export class SamplingState {
-  /**
-   * 
-   * @param {SkeletonInstance} skeleton 
-   */
-  constructor(skeleton) {
+@ccclass('cc.AnimationClip')
+export default class AnimationClip extends Asset {
     /**
-     * @type {SkeletonInstance}
-     * @ignore
+     * @type {AnimationFrame[]}
      */
-    this._skeleton = skeleton;
+    @property
+    _frames = [];
 
     /**
-     * @type {SamplingStateJointState[]}
+     * @type {number}
      */
-    this._jointStates = new Array(skeleton._nodes.length);
-    for (let i = 0; i < this._jointStates.length; ++i)
-      this._jointStates[i] = new SamplingStateJointState(skeleton._nodes[i]);
-  }
-
-  get skeleton() {
-    return this._skeleton;
-  }
-
-  /**
-   * Resets this state to get sampling start.
-   */
-  reset() {
-    for (let i = 0; i < this._skeleton._nodes.length; ++i)
-      this._jointStates[i].reset();
-  }
-
-  /**
-   * Updates the sampling result.
-   */
-  apply() {
-    for (let i = 0; i < this._jointStates.length; ++i)
-      this._jointStates[i].apply();
-    this._skeleton.updateMatrices();
-  }
+    @property
+    _length = 0.0;
+
+    get frames() {
+        return this._frames;
+    }
+
+    get length() {
+        return this._length;
+    }
 }
+cc.AnimationClip = AnimationClip;
+
+export class AnimationTarget {
+    constructor() {
+        /**
+         * @type {Node[]}
+         */
+        this._nodes = [];
+
+        /**
+         * @type {Map<string, number>}
+         */
+        this._pathMap = new Map();
+    }
 
-class SamplingStateJointState {
-  constructor(joint) {
-    /**
-     * @type {Joint}
-     */
-    this._joint = joint;
+    add(path, node) {
+        if (this._pathMap.has(path)) {
+            return;
+        }
+        const index = this._nodes.length;
+        this._nodes.push(node);
+        this._pathMap.set(path, index);
+    }
+
+    get(path) {
+        return this._pathMap.get(path);
+    }
+
+    get nodes() {
+        return this._nodes;
+    }
+
+    getDefaultPosition(nodeIndex) {
+        return this._nodes[nodeIndex].getPosition();
+        //return this._defaultPositions[nodeIndex];
+    }
+
+    getDefaultScale(nodeIndex) {
+        return this._nodes[nodeIndex].getScale();
+        //return this._defaultScales[nodeIndex];
+    }
+
+    getDefaultRotation(nodeIndex) {
+        return this._nodes[nodeIndex].getRotation();
+        //return this._defaultRotations[nodeIndex];
+    }
+
+    _saveDefaultTRS() {
+        this._defaultPositions = new Array(this._nodes.length);
```

Comment:
there are two way for default position, scale and rotation, which one is better?

Reply:
This function currently isn't used. I have commented it out.